docs/examples/data/checkpointing/main.py

```diff
@@ -0,0 +1,221 @@
+"""Single-GPU training example."""
+import logging
+import os
+import shutil
+
+import rich.logging
+import torch
+from torch import Tensor, nn
+from torch.nn import functional as F
+from torch.utils.data import DataLoader, random_split
+from torchvision import transforms
+from torchvision.datasets import CIFAR10
+from torchvision.models import resnet18
+from tqdm import tqdm
+
+
+try:
+    _CHECKPTS_DIR = f"{os.environ['SCRATCH']}/checkpoints"
+except KeyError:
+    _CHECKPTS_DIR = "../checkpoints"
+
+
+def main():
+    training_epochs = 10
+    learning_rate = 5e-4
+    weight_decay = 1e-4
+    batch_size = 128
+    resume_file = f"{_CHECKPTS_DIR}/resnet18_cifar10/checkpoint.pth.tar"
+    start_epoch = 0
+    best_acc = 0
+
+    # Check that the GPU is available
+    assert torch.cuda.is_available() and torch.cuda.device_count() > 0
+    device = torch.device("cuda", 0)
+
+    # Setup logging (optional, but much better than using print statements)
+    logging.basicConfig(
+        level=logging.INFO,
+        handlers=[rich.logging.RichHandler(markup=True)],  # Very pretty, uses the `rich` package.
+    )
+
+    logger = logging.getLogger(__name__)
+
+    # Create a model.
+    model = resnet18(num_classes=10)
+
+    # Resume from a checkpoint
+    if os.path.isfile(resume_file):
+        logger.debug(f"=> loading checkpoint '{resume_file}'")
+        # Map model to be loaded to gpu.
+        checkpoint = torch.load(resume_file, map_location="cuda:0")
+        start_epoch = checkpoint["epoch"]
+        best_acc = checkpoint["best_acc"]
+        # best_acc may be from a checkpoint from a different GPU
+        best_acc = best_acc.to(device)
+        model.load_state_dict(checkpoint["state_dict"])
+        optimizer.load_state_dict(checkpoint["optimizer"])
+        logger.debug(f"=> loaded checkpoint '{resume_file}' (epoch {checkpoint['epoch']})")
```

Comment:
This should also restore the torch/numpy/random RNG states from the saved checkpoint if possible.